primitives/src/parachain.rs

```diff
@@ -78,6 +78,67 @@ pub type ValidatorPair = validator_app::Pair;
 /// so we define it to be the same type as `SessionKey`. In the future it may have different crypto.
 pub type ValidatorSignature = validator_app::Signature;
 
+/// Retriability for a given active para.
+#[derive(Clone, Eq, PartialEq, Encode, Decode)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub enum Retriable {
+	/// Ineligible for retry. This means it's either a parachain which is always scheduled anyway or
+	/// has been removed/swapped.
+	Never,
+	/// Eligible for retry; the associated value is the number of retries that the para already had.
+	WithRetries(u32),
+}
+
+/// Type determining the active set of parachains in current block.
+pub trait ActiveParas {
+	/// Return the active set of parachains in current block. This attempts to keep any IDs in the
+	/// same place between sequential blocks. It is therefore unordered. The second item in the
+	/// tuple is the required collator ID, if any. If `Some`, then it is invalid to include any
+	/// other collator's block.
+	///
+	/// NOTE: The initial implementation simply concatenates the (ordered) set of (permanent)
+	/// parachain IDs with the (unordered) set of parathread IDs selected for this block.
+	fn active_paras() -> Vec<(Id, Option<(CollatorId, Retriable)>)>;
+}
+
+/// Description of how often/when this parachain is scheduled for progression.
+#[derive(Encode, Decode, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub enum Scheduling {
+	/// Scheduled every block.
+	Always,
+	/// Scheduled dynamically (i.e. a parathread).
+	Dynamic,
+}
+
+/// Information regarding a deployed parachain/thread.
+#[derive(Encode, Decode, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct Info {
+	/// Scheduling info.
+	pub scheduling: Scheduling,
+}
+
+/// An `Info` value for a standard leased parachain.
+pub const PARACHAIN_INFO: Info = Info {
+	scheduling: Scheduling::Always,
+};
+
+/// Handler for when two parachains/parathreads get notionally swapped.
+pub trait OnSwap {
+	/// Result describing whether it is possible to swap two parachains. Doesn't mutate state.
+	fn ensure_can_swap(one: Id, other: Id) -> Result<(), &'static str>;
+
+	/// Updates any needed state/references to enact a logical swap of two parachains. Identity,
+	/// code and head_data remain equivalent for all parachains/threads, however other properties
+	/// such as leases, deposits held and thread/chain nature are swapped.
+	///
+	/// May only be called on a state that `ensure_can_swap` has previously returned `Ok` for: if this is
+	/// not the case, the result is undefined. May only return an error if `ensure_can_swap` also returns
+	/// an error.
+	fn on_swap(one: Id, other: Id) -> Result<(), &'static str>;
```

Comment:
it's also strange to me to call this `on_swap`, when it's rather asking the trait to `do_swap` or `perform_swap` or just plain old `swap` two parachains. We usually use `OnX` traits as event handlers.

Reply:
it is an event: two paras are being swapped. there may be things that need to be done outside of the entity that is actually managing the swap, so this is a hook to do those things.

Reply:
typically an event handler is something that is informed when an event has occurred, but cannot determine whether that event is going to occur or not

Reply:
the `on_swap` (i.e. handler) portion of this trait cannot determine if it's going to happen.

Reply:
yes, that's the interface as written. However, the result is undefined if `can_swap` has not returned `Ok(())`. That means that whatever carries out the event has three options:
  - Execute `on_swap` for all event handlers without asking about `can_swap`, risking triggering undefined behavior
  - Execute `can_swap` for all event handlers after applying the event, and simply don't trigger `on_swap` for those which cannot swap. This is just strange, since the event handler could just ignore internally, if that's the desired goal.
  - Execute `can_swap` for all event handlers before applying the event, and if any cannot handle it, do not apply the event. That is not a proper event handler, because it alters the control flow of the notifier.

Reply:
Will rename `SwapAuxilliary`?

Reply:
done.